bk-nodes/bk_db.py

```diff
@@ -0,0 +1,241 @@
+
+import MySQLdb
+import config
+import dateutil.parser
+
+
+table_fields = {}
+table_fields_index ={}
+
+
+class BeekeeperDB():
+    def __init__ ( self , retries=60) :
+
+        if not config.mysql_host:
+            raise Exception("mysql_host is not defined")
+        
+        count = 0
+        while True:
+            try:
+                self.db=MySQLdb.connect(host=config.mysql_host,user=config.mysql_user,
+                  passwd=config.mysql_password,db=config.mysql_db)
+            except Exception as e: # pragma: no cover
+                if count > retries:
+                    raise
+                print(f'Could not connnect to database, error={e}, retry in 2 seconds', file=sys.stderr)
+                time.sleep(2)
+                count += 1
+                continue
+            break
+
+        self.cur=self.db.cursor()
+        return
+
+
+    def nodes_log_add(self, logData):
+        #node_id, table_name, operation, field_name, new_value, source, effective_time=None
+        fields = '`node_id`, `table_name`, `operation`, `field_name`, `new_value`, `source`, `effective_time`'
+        values_s = '%s, %s, %s, %s, %s, %s, %s'
+        #values = [node_id, table_name, operation, field_name, new_value, source]
+        
+        smallest_timestamp = None 
+        values = []
+        for log in logData:
+            effective_time =  dateutil.parser.parse(log["effective_time"])
+            
+            if (not smallest_timestamp) or (effective_time < smallest_timestamp):
+                smallest_timestamp = effective_time
+
+            value_tuple =(log["node_id"], log["table_name"], log["operation"], log["field_name"], log["new_value"], log["source"], effective_time)
+            values.append(value_tuple)
+
+
+
+
+        self.cur.execute("LOCK TABLES `nodes_log` WRITE")
+
+        stmt = f'INSERT INTO `nodes_log` ( {fields} ) VALUES ({values_s})'
+        #debug_stmt = stmt
+        for tup in values:
+            debug_stmt = stmt
+            for i in tup:
+                debug_stmt = debug_stmt.replace("%s", f'"{i}"', 1)
+            print(f'debug_stmt: {debug_stmt}', flush=True)
+
+        #print(f'debug_stmt: {debug_stmt}', file=sys.stderr)
+        
+        self.cur.executemany(stmt, values )
+        #self.cur.execute("UNLOCK TABLES")
+        self.db.commit()
+
+        try:
+
+            self.replay_log(replay_from_timestamp = smallest_timestamp)
+        except Exception as e:
+            raise Exception("Log replay failed: "+str(e))
```

Comment:
i dont know how to recover from this?

the `node_history` is now messed up.  should we assume that the inserted entry into `node_log` was the reason for the failure and remove it and re-call `replay_log()`?

in theory the it should succeed as we are restoring the `node_log` table to the previous state when the last time the call to `replay_log()` was successful?

Reply:
Ideally the insertion into the log and the history replay should be one single(atomic) transaction. I am going to take a look at that, I am not sure how easy that would be.

Reply:
okay sounds good.  since the /state call references the node_history table... if its messed up it could be bad.

yes, it is a cache and can be re-created (in theory).  but if the last entry into the node_log is causing the node_history to fail (for whatever reason).  we want to identify it now, instead of down the road.

yes this exception should be super rare.  but if it does occur, i think raisign an exception is a good idea.  but maybe restoring the DB back to the way it was would be nice.

anywaay... u said you would think about it. :)